### DISCUSSION TITLE
bloblang method to URL encode string

### DISCUSSION THREAD

**Discussion by gramian** [redpanda-data/connect#1615] [Q&A] [answered]
Dear Developers, Dear Community,

I am looking for a way to [URL encode](https://en.wikipedia.org/wiki/Percent-encoding) a string via bloblang. However, I could only find [`escape_url_query`](https://www.benthos.dev/docs/guides/bloblang/methods#escape_url_query). Yet, I don't have a query, I need to pass an URL encoded string in a REST API. Is this functionality missing, or did I overlook a method for this purpose?

Thank you

**Comment by mihaitodor** [comment, marked as answer]
Hey @gramian. I think the name of `escape_url_query` might be a bit misleading, but the docs do state that it "Escapes a string so that it can be safely placed within a URL query". Under the hood, it simply calls [`url.QueryEscape`](https://pkg.go.dev/net/url#QueryEscape), which, as far as I can tell, performs the percent encoding that you're looking for. I tested some examples from Mozilla's [`encodeURI()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI) docs quickly via `benthos blobl server` and got the same results.

**Reply by gramian** [reply]
That is cool. I asked because the example given in [escape_url_query](https://www.benthos.dev/docs/guides/bloblang/methods/#escape_url_query) suggested to me that spaces are not translated to `%20` but to `+`.

**Reply by mihaitodor** [reply]
Ugh, indeed, it does look like there is an edge case related to spaces... Thanks for pointing it out! There's a bug report on the Go repo [here](https://github.com/golang/go/issues/4013) regarding this, which I guess led them to also expose [`url.PathEscape()`](https://pkg.go.dev/net/url#PathEscape). As far as I can tell, `url.PathEscape()` encodes spaces to `%20`, while `url.QueryEscape()` encodes them to `+`. Guess we'll have to add a separate method for percent encoding after all. I opened an issue for it in #1617.

In the mean time, I think you could achieve what you want by using [`replace_all_many()`](https://www.benthos.dev/docs/guides/bloblang/methods/#replace_all_many) instead of `escape_url_query()`, but there might be some nasty corner cases...
